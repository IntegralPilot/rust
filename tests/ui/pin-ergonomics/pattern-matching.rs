//@ revisions: pin_ergonomics normal
//@ edition:2024
//@[pin_ergonomics] check-pass
#![cfg_attr(pin_ergonomics, feature(pin_ergonomics))]
#![feature(if_let_guard, negative_impls)]
#![allow(incomplete_features)]

use std::pin::Pin;

// This test verifies that a `&pin mut T` can be projected to a pinned
// reference field `&pin mut T.U`.
// FIXME(pin_ergonomics): it is unsound to project `&pin mut T` to
// `&pin mut T.U` when `U: !Unpin` but `T: Unpin`, or when there exists
// `impl Drop for T` that takes a `&mut self` receiver.

struct Foo<T, U> {
    x: T,
    y: U,
}

enum Bar<T, U> {
    Foo(T, U),
    Bar { x: T, y: U },
}

trait IsPinMut {}
trait IsPinConst {}
impl<T: ?Sized> IsPinMut for Pin<&mut T> {}
impl<T: ?Sized> IsPinConst for Pin<&T> {}

fn assert_pin_mut<T: IsPinMut>(_: T) {}
fn assert_pin_const<T: IsPinConst>(_: T) {}

fn foo<T: Unpin, U: Unpin>(foo_mut: Pin<&mut Foo<T, U>>, foo_const: Pin<&Foo<T, U>>) {
    let Foo { x, y } = foo_mut;
    //[normal]~^ ERROR mismatched types
    assert_pin_mut(x);
    assert_pin_mut(y);

    let Foo { x, y } = foo_const;
    //[normal]~^ ERROR mismatched types
    assert_pin_const(x);
    assert_pin_const(y);
}

fn bar<T: Unpin, U: Unpin>(bar_mut: Pin<&mut Bar<T, U>>, bar_const: Pin<&Bar<T, U>>) {
    match bar_mut {
        Bar::Foo(x, y) => {
            //[normal]~^ ERROR mismatched types
            assert_pin_mut(x);
            assert_pin_mut(y);
        }
        _ if let Bar::Bar { x, y } = bar_mut => {
            //[normal]~^ ERROR mismatched types
            assert_pin_mut(x);
            assert_pin_mut(y);
        }
        _ => {}
    }
    match bar_const {
        Bar::Bar { x, y } => {
            //[normal]~^ ERROR mismatched types
            assert_pin_const(x);
            assert_pin_const(y);
        }
        _ if let Bar::Foo(x, y) = bar_const => {
            //[normal]~^ ERROR mismatched types
            assert_pin_const(x);
            assert_pin_const(y);
        }
        _ => {}
    }
}

fn pin_mut_tuple<T, U>(foo_mut: Pin<&mut (Foo<T, U>,)>, foo_const: Pin<&(Foo<T, U>,)>) {
    let (Foo { x, y },) = foo_mut;
    //[normal]~^ ERROR mismatched types
    assert_pin_mut(x);
    assert_pin_mut(y);
    let (Foo { x, y },) = foo_const;
    //[normal]~^ ERROR mismatched types
    assert_pin_const(x);
    assert_pin_const(y);
}

fn pin_mut_array<T, U>(foo_mut: Pin<&mut [Foo<T, U>; 1]>, foo_const: Pin<&[Foo<T, U>; 1]>) {
    let [Foo { x, y }] = foo_mut;
    //[normal]~^ ERROR expected an array or slice, found `Pin<&mut [Foo<T, U>; 1]>`
    assert_pin_mut(x);
    assert_pin_mut(y);
    let [Foo { x, y }] = foo_const;
    //[normal]~^ ERROR expected an array or slice, found `Pin<&[Foo<T, U>; 1]>`
    assert_pin_const(x);
    assert_pin_const(y);
}

fn pin_mut_slice<T, U>(foo_mut: Pin<&mut [Foo<T, U>]>, foo_const: Pin<&[Foo<T, U>]>) {
    if let [Foo { x, y }] = foo_mut {
        //[normal]~^ ERROR expected an array or slice, found `Pin<&mut [Foo<T, U>]>`
        assert_pin_mut(x);
        assert_pin_mut(y);
    }
    if let [Foo { x, y }] = foo_const {
        //[normal]~^ ERROR expected an array or slice, found `Pin<&[Foo<T, U>]>`
        assert_pin_const(x);
        assert_pin_const(y);
    }
}

fn main() {}
