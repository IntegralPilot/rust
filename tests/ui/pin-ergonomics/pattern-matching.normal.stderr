error[E0308]: mismatched types
  --> $DIR/pattern-matching.rs:35:9
   |
LL |     let Foo { x, y } = foo_mut;
   |         ^^^^^^^^^^^^   ------- this expression has type `Pin<&mut Foo<T, U>>`
   |         |
   |         expected `Pin<&mut Foo<T, U>>`, found `Foo<_, _>`
   |
   = note: expected struct `Pin<&mut Foo<T, U>>`
              found struct `Foo<_, _>`
help: consider dereferencing to access the inner value using the Deref trait
   |
LL |     let Foo { x, y } = *foo_mut;
   |                        +

error[E0308]: mismatched types
  --> $DIR/pattern-matching.rs:40:9
   |
LL |     let Foo { x, y } = foo_const;
   |         ^^^^^^^^^^^^   --------- this expression has type `Pin<&Foo<T, U>>`
   |         |
   |         expected `Pin<&Foo<T, U>>`, found `Foo<_, _>`
   |
   = note: expected struct `Pin<&Foo<T, U>>`
              found struct `Foo<_, _>`
help: consider dereferencing to access the inner value using the Deref trait
   |
LL |     let Foo { x, y } = *foo_const;
   |                        +

error[E0308]: mismatched types
  --> $DIR/pattern-matching.rs:48:9
   |
LL |     match bar_mut {
   |           ------- this expression has type `Pin<&mut Bar<T, U>>`
LL |         Bar::Foo(x, y) => {
   |         ^^^^^^^^^^^^^^ expected `Pin<&mut Bar<T, U>>`, found `Bar<_, _>`
   |
   = note: expected struct `Pin<&mut Bar<T, U>>`
                found enum `Bar<_, _>`
help: consider dereferencing to access the inner value using the Deref trait
   |
LL |     match *bar_mut {
   |           +

error[E0308]: mismatched types
  --> $DIR/pattern-matching.rs:53:18
   |
LL |         _ if let Bar::Bar { x, y } = bar_mut => {
   |                  ^^^^^^^^^^^^^^^^^   ------- this expression has type `Pin<&mut Bar<T, U>>`
   |                  |
   |                  expected `Pin<&mut Bar<T, U>>`, found `Bar<_, _>`
   |
   = note: expected struct `Pin<&mut Bar<T, U>>`
                found enum `Bar<_, _>`
help: consider dereferencing to access the inner value using the Deref trait
   |
LL |         _ if let Bar::Bar { x, y } = *bar_mut => {
   |                                      +

error[E0308]: mismatched types
  --> $DIR/pattern-matching.rs:61:9
   |
LL |     match bar_const {
   |           --------- this expression has type `Pin<&Bar<T, U>>`
LL |         Bar::Bar { x, y } => {
   |         ^^^^^^^^^^^^^^^^^ expected `Pin<&Bar<T, U>>`, found `Bar<_, _>`
   |
   = note: expected struct `Pin<&Bar<T, U>>`
                found enum `Bar<_, _>`
help: consider dereferencing to access the inner value using the Deref trait
   |
LL |     match *bar_const {
   |           +

error[E0308]: mismatched types
  --> $DIR/pattern-matching.rs:66:18
   |
LL |         _ if let Bar::Foo(x, y) = bar_const => {
   |                  ^^^^^^^^^^^^^^   --------- this expression has type `Pin<&Bar<T, U>>`
   |                  |
   |                  expected `Pin<&Bar<T, U>>`, found `Bar<_, _>`
   |
   = note: expected struct `Pin<&Bar<T, U>>`
                found enum `Bar<_, _>`
help: consider dereferencing to access the inner value using the Deref trait
   |
LL |         _ if let Bar::Foo(x, y) = *bar_const => {
   |                                   +

error[E0308]: mismatched types
  --> $DIR/pattern-matching.rs:76:9
   |
LL |     let (Foo { x, y },) = foo_mut;
   |         ^^^^^^^^^^^^^^^   ------- this expression has type `Pin<&mut (Foo<T, U>,)>`
   |         |
   |         expected `Pin<&mut (Foo<T, U>,)>`, found `(_,)`
   |
   = note: expected struct `Pin<&mut (Foo<T, U>,)>`
               found tuple `(_,)`
help: consider dereferencing to access the inner value using the Deref trait
   |
LL |     let (Foo { x, y },) = *foo_mut;
   |                           +

error[E0308]: mismatched types
  --> $DIR/pattern-matching.rs:80:9
   |
LL |     let (Foo { x, y },) = foo_const;
   |         ^^^^^^^^^^^^^^^   --------- this expression has type `Pin<&(Foo<T, U>,)>`
   |         |
   |         expected `Pin<&(Foo<T, U>,)>`, found `(_,)`
   |
   = note: expected struct `Pin<&(Foo<T, U>,)>`
               found tuple `(_,)`
help: consider dereferencing to access the inner value using the Deref trait
   |
LL |     let (Foo { x, y },) = *foo_const;
   |                           +

error[E0529]: expected an array or slice, found `Pin<&mut [Foo<T, U>; 1]>`
  --> $DIR/pattern-matching.rs:87:9
   |
LL |     let [Foo { x, y }] = foo_mut;
   |         ^^^^^^^^^^^^^^ pattern cannot match with input type `Pin<&mut [Foo<T, U>; 1]>`

error[E0529]: expected an array or slice, found `Pin<&[Foo<T, U>; 1]>`
  --> $DIR/pattern-matching.rs:91:9
   |
LL |     let [Foo { x, y }] = foo_const;
   |         ^^^^^^^^^^^^^^ pattern cannot match with input type `Pin<&[Foo<T, U>; 1]>`

error[E0529]: expected an array or slice, found `Pin<&mut [Foo<T, U>]>`
  --> $DIR/pattern-matching.rs:98:12
   |
LL |     if let [Foo { x, y }] = foo_mut {
   |            ^^^^^^^^^^^^^^ pattern cannot match with input type `Pin<&mut [Foo<T, U>]>`

error[E0529]: expected an array or slice, found `Pin<&[Foo<T, U>]>`
  --> $DIR/pattern-matching.rs:103:12
   |
LL |     if let [Foo { x, y }] = foo_const {
   |            ^^^^^^^^^^^^^^ pattern cannot match with input type `Pin<&[Foo<T, U>]>`

error: aborting due to 12 previous errors

Some errors have detailed explanations: E0308, E0529.
For more information about an error, try `rustc --explain E0308`.
